server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java

```diff
@@ -0,0 +1,300 @@
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.druid.java.util.common.IAE;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class keeps a bounded list of segment updates made on the server such as adding/dropping segments.
+ *
+ * Clients call addSegmentChangeRequest(DataSegmentChangeRequest) or addSegmentChangeRequests(List<DataSegmentChangeRequest>)
+ * to add segment updates.
+ *
+ * Clients call ListenableFuture<SegmentChangeRequestsSnapshot> getRequestsSince(final Counter counter) to get segment
+ * updates since given counter.
+ */
+public class SegmentChangeRequestHistory
+{
+  private static int MAX_SIZE = 1000;
+
+  private final int maxSize;
+
+  @GuardedBy("this")
+  private final List<Holder> changes;
+
+  @VisibleForTesting
+  @GuardedBy("waitingFutures")
+  final LinkedHashMap<CustomSettableFuture, Counter> waitingFutures;
+
+  private final ExecutorService singleThreadedExecutor;
+  private final Runnable resolveWaitingFuturesRunnable;
+
+  public SegmentChangeRequestHistory()
+  {
+    this(MAX_SIZE);
+  }
+
+  @VisibleForTesting
+  SegmentChangeRequestHistory(int maxSize)
+  {
+    this.maxSize = maxSize;
+    this.changes = new ArrayList<>(maxSize);
+
+    this.waitingFutures = new LinkedHashMap<>();
+
+    this.resolveWaitingFuturesRunnable = new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        resolveWaitingFutures();
+      }
+    };
+
+    this.singleThreadedExecutor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setDaemon(true)
+                                  .setNameFormat(
+                                      "SegmentChangeRequestHistory"
+                                  )
+                                  .build()
+    );
+  }
+
+
+
+  /**
+   * Add batch of segment changes update.
+   */
+  public synchronized void addSegmentChangeRequests(List<DataSegmentChangeRequest> requests)
+  {
+    for (DataSegmentChangeRequest request : requests) {
+      if (changes.size() >= maxSize) {
+        changes.remove(0);
+      }
+      changes.add(new Holder(request, getLastCounter().inc()));
+    }
+
+    singleThreadedExecutor.execute(resolveWaitingFuturesRunnable);
+  }
+
+  /**
+   * Add single segment change update.
+   */
+  public synchronized void addSegmentChangeRequest(DataSegmentChangeRequest request)
+  {
+    addSegmentChangeRequests(ImmutableList.of(request));
+  }
+
+  /**
+   * Returns a Future that , on completion, returns list of segment updates and associated counter.
+   * If there are no update since given counter then Future completion waits till an updates is provided.
+   *
+   * If counter is very older than max number of changes maintained than IAE exception is thrown.
```

Comment:
I'm not sure that an exception is correct.  It might be nicer to have the `SegmentChangeRequestsSnapshot` have some indication that it is empty or invalid because the counter was too old.